server/applicationset/applicationset.go

```diff
@@ -0,0 +1,366 @@
+	if err != nil {
+		return nil, fmt.Errorf("error updating ApplicationSets: %w", err)
+	}
+	return updated, nil
+}
+
+func mergeStringMaps(items ...map[string]string) map[string]string {
+	res := make(map[string]string)
+	for _, m := range items {
+		if m == nil {
+			continue
+		}
+		for k, v := range m {
+			res[k] = v
+		}
+	}
+	return res
+}
+
+func (s *Server) updateAppSet(appset *v1alpha1.ApplicationSet, newAppset *v1alpha1.ApplicationSet, ctx context.Context, merge bool) (*v1alpha1.ApplicationSet, error) {
+
+	if appset != nil && appset.Spec.Template.Spec.Project != newAppset.Spec.Template.Spec.Project {
+		// When changing projects, caller must have applicationset create, update & delete privileges in new project
+		// NOTE: the update check was already verified in the caller to this function
+		if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceApplicationSets, rbacpolicy.ActionCreate, apputil.AppSetRBACName(newAppset)); err != nil {
+			return nil, err
+		}
+		// They also need 'update' privileges in the old project
+		if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceApplicationSets, rbacpolicy.ActionUpdate, apputil.AppSetRBACName(appset)); err != nil {
+			return nil, err
+		}
+		// They also need 'delete' privileges in the old project
+		if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceApplicationSets, rbacpolicy.ActionDelete, apputil.AppSetRBACName(appset)); err != nil {
+			return nil, err
+		}
+	}
+
+	for i := 0; i < 10; i++ {
+		appset.Spec = newAppset.Spec
+		if merge {
+			appset.Labels = mergeStringMaps(appset.Labels, newAppset.Labels)
+			appset.Annotations = mergeStringMaps(appset.Annotations, newAppset.Annotations)
+		} else {
+			appset.Labels = newAppset.Labels
+			appset.Annotations = newAppset.Annotations
+		}
+
+		res, err := s.appclientset.ArgoprojV1alpha1().ApplicationSets(s.ns).Update(ctx, appset, metav1.UpdateOptions{})
+		if err == nil {
+			s.logAppSetEvent(appset, ctx, argo.EventReasonResourceUpdated, "updated ApplicationSets spec")
+			s.waitSync(res)
+			return res, nil
+		}
+		if !apierr.IsConflict(err) {
+			return nil, err
+		}
+
+		appset, err = s.appclientset.ArgoprojV1alpha1().ApplicationSets(s.ns).Get(ctx, newAppset.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("error getting ApplicationSets: %w", err)
+		}
+	}
+	return nil, status.Errorf(codes.Internal, "Failed to update ApplicationSets. Too many conflicts")
+}
+
+func (s *Server) Delete(ctx context.Context, q *applicationset.ApplicationSetDeleteRequest) (*applicationset.ApplicationSetResponse, error) {
+
+	appset, err := s.appclientset.ArgoprojV1alpha1().ApplicationSets(s.ns).Get(ctx, q.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error getting ApplicationSets: %w", err)
+	}
+
+	s.projectLock.RLock(appset.Spec.Template.Spec.Project)
+	defer s.projectLock.RUnlock(appset.Spec.Template.Spec.Project)
```

Comment:
I think the lock could be moved below the RBAC check

Reply:
Yes, that makes sense